File: days/day_03/solutions/solution.ts
import * as fs from 'fs';
import * as path from 'path';

// Find the largest N-digit number by selecting N digits in order from bank
function findLargestSubsequence(bank: string, n: number): number {
    if (bank.length < n) {
        return 0;
    }
    
    const result: string[] = [];
    let start = 0;
    const bankLen = bank.length;
    
    for (let i = 0; i < n; i++) {
        const remainingNeeded = n - i - 1;
        const end = bankLen - remainingNeeded;
        
        let maxDigit = bank[start];
        let maxPos = start;
        for (let j = start + 1; j < end; j++) {
            if (bank[j] > maxDigit) {
                maxDigit = bank[j];
                maxPos = j;
            }
        }
        
        result.push(maxDigit);
        start = maxPos + 1;
    }
    
    return parseInt(result.join(''), 10);
}

function solve(inputData: string): [string, string] {
    const lines = inputData.trim().split('\n');
    
    let part1Sum = 0;
    let part2Sum = 0;
    
    for (const line of lines) {
        const bank = line.trim();
        if (bank === '') {
            continue;
        }
        
        part1Sum += findLargestSubsequence(bank, 2);
        
        if (bank.length >= 12) {
            part2Sum += findLargestSubsequence(bank, 12);
        }
    }
    
    return [part1Sum.toString(), part2Sum.toString()];
}

function main(): void {
    const inputPath = path.join(__dirname, '../data/input.txt');
    const data = fs.readFileSync(inputPath, 'utf-8');
    const [part1, part2] = solve(data);
    console.log(`Part 1: ${part1}`);
    console.log(`Part 2: ${part2}`);
}

main();

File: days/day_03/solutions/solution.rs
use std::fs;

// Find the largest N-digit number by selecting N digits in order from bank
fn find_largest_subsequence(bank: &str, n: usize) -> u64 {
    let bank_len = bank.len();
    if bank_len < n {
        return 0;
    }
    
    let mut result = Vec::new();
    let mut start = 0;
    
    for i in 0..n {
        let remaining_needed = n - i - 1;
        let end = bank_len - remaining_needed;
        
        let mut max_digit = bank.chars().nth(start).unwrap();
        let mut max_pos = start;
        for j in (start + 1)..end {
            let ch = bank.chars().nth(j).unwrap();
            if ch > max_digit {
                max_digit = ch;
                max_pos = j;
            }
        }
        
        result.push(max_digit);
        start = max_pos + 1;
    }
    
    result.iter().collect::<String>().parse::<u64>().unwrap_or(0)
}

fn solve(input_data: &str) -> (String, String) {
    let lines: Vec<&str> = input_data.trim().lines().collect();
    
    let mut part1_sum: u64 = 0;
    let mut part2_sum: u64 = 0;
    
    for line in lines {
        let bank = line.trim();
        if bank.is_empty() {
            continue;
        }
        
        part1_sum += find_largest_subsequence(bank, 2);
        
        if bank.len() >= 12 {
            part2_sum += find_largest_subsequence(bank, 12);
        }
    }
    
    (part1_sum.to_string(), part2_sum.to_string())
}

fn main() {
    let data = fs::read_to_string("../data/input.txt").expect("Error reading file");
    let (part1, part2) = solve(&data);
    println!("Part 1: {}", part1);
    println!("Part 2: {}", part2);
}
